[RFR][Feature][OSF-5939]Sort files by Dates in project(overview) page and project files page

TomBaxter opened the pull request:
## Purpose:
[OSF-5939](https://openscience.atlassian.net/browse/OSF-5939)
In project overview page and project file view page...
- Add last modified date to file info.
- Allow sorting by modified date

## Bonus
GitHub files now show their file size in project files view
Dataverse Raven logging should now work

## Changes:
Update website/addons/dataverse/static/dataverseFangornConfig.js
- Require raven-js to resolve unknown name
- Add columns size, downloads, and modified to allow for sorting by last modified date

Update website/addons/github/static/githubFangornConfig.js
- Minor clean up
- Add columns size, downloads, and modified to allow for sorting by last modified date

Update  website/static/js/fangorn.js 
- Require moment-js for handling date/time
- Add  function _fangornModifiedColumn to normalize modified date using moment-js
- Update function _fangornResolveRows to fill in size, download, and modified for all files
- Allow for sorting by last modified date

Update website/static/js/pages/project-dashboard-page.js
- Update $(document).ready(function () to show sortable modified column

## Side effects
None

## Caveats
WB does not provide file last modified dates for the following providers...
Dataverse
FigShare
GitHub

WB does not provide file sizes for the Dataverse provider

WB only provides download counts for the OSFStorage provider


[#OSF-5939]

mfraezz commented:
Looks good on the project dashboard:

<img width="530" alt="screen shot 2016-06-10 at 16 50 17" src="https://cloud.githubusercontent.com/assets/5659262/15978416/d6d17748-2f2b-11e6-8d75-80a4b2ae7774.png">

However, there's a 🐛 with the files page. The requests and responses seem to be going out correctly, but the UI is broken:

<img width="1009" alt="screen shot 2016-06-10 at 16 52 16" src="https://cloud.githubusercontent.com/assets/5659262/15978453/0eeeeba6-2f2c-11e6-9c78-d1ce3dc89236.png">

mfraezz commented on website/addons/dataverse/static/dataverseFangornConfig.js:
```diff
@@ -254,25 +255,36 @@ function _fangornDataverseTitle(item, col) {
     });
-
-    if (tb.options.placement === 'project-files') {
+    if(tb.options.placement === 'project-files') {
```
This line and the other conditional below should both have a space between `if` and `(`, as per [cosdev](http://cosdev.readthedocs.io/en/latest/style_guides/javascript.html)

mfraezz commented:
Pass complete, please also resolve the merge conflicts. Back to you @TomBaxter 🐠

TomBaxter commented:
File view fixed for non-existant item.data.extra.downloads effecting folders in OSFStorage.
Removed console statements.
Fixed no space between "if" and "(".
Fixed merge conflict.
Ready for further review.

mfraezz commented on website/addons/github/static/githubFangornConfig.js:
```diff
@@ -340,11 +337,27 @@ function _fangornColumns (item) {
+        });
+    }
+
+    if(tb.options.placement === 'project-files') {
```
This looks like the same conditional as above; could you combine the `'downloads'` and `'size'` data to the same `.push` call?

TomBaxter replied:
Done.

mfraezz commented:
Minor: I'm getting a `TypeError: item is undefined` from [here](https://github.com/CenterForOpenScience/osf.io/blob/9bb285a705f106dd25197dadb761c2f1790fa9ca/website/static/js/fangorn.js#L2374), but it doesn't seem to be causing any behavior issues. This happens when expanding/collapsing folders (including addon root folders), but only when no rows are 'highlighted'. Confirmed that previously documented :bug: was fixed.

<img width="734" alt="screen shot 2016-06-14 at 14 11 46" src="https://cloud.githubusercontent.com/assets/5659262/16054174/024ae52e-323a-11e6-8bac-4b8e9434f72d.png">

TomBaxter commented:
The TypeError was introduced with [PR#5835](https://github.com/CenterForOpenScience/osf.io/pull/5835). Looking into a fix now.

erinspace commented:
@TomBaxter #5835 should have *fixed* that typeError on expanding and contracting

TomBaxter commented:
@erinspace OK, I'll double check.

TomBaxter commented:
@erinspace You right, sorry about that. Apparently I have dyslexia, but with red and green.

mfraezz commented on website/addons/dataverse/static/dataverseFangornConfig.js:
```diff
@@ -254,25 +255,36 @@ function _fangornDataverseTitle(item, col) {
+            custom : function() {return m('');}
+        });
     }
+    if (tb.options.placement === 'project-files') {
```
Ditto [this comment](https://github.com/CenterForOpenScience/osf.io/pull/5817/files#r67015172)

>Minor: this looks like the same conditional as above; could you combine the 'downloads' and 'size' data to the same .push call?

Perhaps also consider DRYing this code up, but that's not a blocker.

TomBaxter replied:
Done and pushed.

sloria commented on website/static/js/fangorn.js:
```diff
@@ -10,6 +10,7 @@ var m = require('mithril');
 var URI = require('URIjs');
 var Raven = require('raven-js');
 var Treebeard = require('treebeard');
+var Moment = require('moment');
```
Minor: By convention, we do not capitalize `moment`.

TomBaxter replied:
Done.

sloria commented on website/static/js/fangorn.js:
```diff
@@ -1270,6 +1271,30 @@ function _fangornTitleColumn(item, col) {
+        return m(
+            'span',
+            // "new Date" required for non-ISO date formats
+            new Moment(new Date(item.data.modified)).format('YYYY-MM-DD hh:mm A')
```
From the docs, it doesn't appear correct to do `new Moment`. `moment(new Date(...)).format(...)` should work. Can you verify?

TomBaxter replied:
That's how I did it originally but I got a jshint error from Travis without the "new"

sloria commented on website/static/js/fangorn.js:
```diff
@@ -1270,6 +1271,30 @@ function _fangornTitleColumn(item, col) {
+    if (item.data.isAddonRoot && item.connected === false) { // as opposed to undefined, avoids unnecessary setting of this value
+        return _connectCheckTemplate.call(this, item);
+    }
+    var assert = require('assert');
```
This should be removed.

TomBaxter replied:
Done.

sloria commented on website/static/js/fangorn.js:
```diff
@@ -1340,29 +1365,33 @@ function _fangornResolveRows(item) {
+            data : 'downloads',
+            sortInclude : false,
+            filter : false,
+            custom: function() { return item.data.extra && item.data.extra.downloads ? item.data.extra.downloads.toString() : ''; }
```
The preferred way to get deeply nested attributes is to use `lodash.get`.

```js
... return lodashGet(item, 'data.extra.downloads') ? ...
```

TomBaxter replied:
Done, and nifty.

sloria commented:
Pass finished.

sloria commented:
Looks good. 🐙
